racket/collects/racket/file.rkt

```diff
@@ -134,100 +139,410 @@
+;; The entry point for the variants based on `format` (no `*`) is
+;; `do-make-temporary-file/directory:format`; for the `*` variants,
+;; `do-make-temporary-file/directory:bytes-append`.
+;; Each entry point tail-calls `do-make-temporary-file/directory:check-make-name`,
+;; which tries building a path from the template or prefix+suffix and reports
+;; various possible errors with the result.
+;; If the tests pass, it tail-calls `internal-make-temporary-file/directory`.
+
+(define (internal-make-temporary-file/directory who copy-from base-dir make-name)
+  (define tmpdir (find-system-path 'temp-dir))
+  (let loop ([s (current-seconds)]
+             [ms (inexact->exact (truncate (current-inexact-milliseconds)))]
+             [tries 0])
+    (define pth
+      (path->complete-path
+       (let ([n (make-name
+                 ;; docs promise argument will be a string containing only digits
+                 (format "~a~a" s ms))])
+         (cond [base-dir (build-path base-dir n)]
+               [(relative-path? n) (build-path tmpdir n)]
+               [else n]))))
+    (with-handlers ([(lambda (exn)
+                       (or (exn:fail:filesystem:exists? exn)
+                           (and (exn:fail:filesystem:errno? exn)
+                                (let ([errno (exn:fail:filesystem:errno-errno exn)])
+                                  (and (eq? 'windows (cdr errno))
+                                       (eqv? (car errno) 5) ; ERROR_ACCESS_DENIED
+                                       ;; On Windows, if the target path refers to a file
+                                       ;; that has been deleted but is still open
+                                       ;; somewhere, then an access-denied error is reported
+                                       ;; instead of a file-exists error; there appears
+                                       ;; to be no way to detect that it was really a
+                                       ;; file-still-exists error. Try again for a while.
+                                       ;; There's still a small chance that this will
+                                       ;; fail, but it's vanishingly small at 32 tries.
+                                       ;; If ERROR_ACCESS_DENIED really is the right
+                                       ;; error (e.g., because the target directory is not
+                                       ;; writable), we'll take longer to get there.
+                                       (tries . < . 32))))))
+                     (lambda (x)
+                       ;; try again with a new name
+                       (loop (- s (random 10))
+                             (+ ms (random 10))
+                             (add1 tries)))])
+      (if copy-from
+          (if (eq? copy-from 'directory)
+              (make-directory pth)
+              (copy-file copy-from pth))
+          (close-output-port (open-output-file pth)))
+      pth)))
```

Comment:
As I look at this again, it seems there's no limit on the number of tries other than for `ERROR_ACCESS_DENIED` on Windows. It seems like it might be better to have some very high limit on tries in general—100? 1000?—so that, if something is going badly wrong, we eventually raise an exception, rather than looping forever.